### ISSUE TITLE
Support middleware that delays processing

### ISSUE THREAD

**Issue by legionth** [legionth/php-http-server-react#14]
@clue had a question about delays of processing in an middlware. Can be found in PR #12 

> I'd like to implement a middleware that delays processing by 2.0s for certain users, is this possible already?

The ToDo's in thise Issues are to validate if it's possible and to create a support for this behavior.

**Comment by clue** [comment]
Obviously, a middleware is not allowed to `sleep(2)`, but how about `$loop->addTimer(2.0, $fn);`?

I suppose the middleware will have to return a promise in this case? How does this fit with the rest of the API design?

**Comment by legionth** [comment]
It is no problem when your middleware returns a promise. But you have to consider that your other middlewares in the middleware chain may have to deal with this promise too (just a friendly reminder 💡 ).

I guess I have to update the documentation about this.

**Comment by clue** [comment]
Does this mean *each and every* middleware always has to check for both direct return values and exceptions as well as promises? If so, does it make sense to unify this interface and perhaps implement this as an adapter internally, so that each consumer can rely on a single interface?

**Comment by legionth** [comment]
Good point. I could change the adapter a bit, so every `$next($request)` would return a promise.

**Comment by clue** [comment]
That would be awesome! This could also take care of validating return values and ensuring the interface is always met :+1:
